README.md
# hc-test-util-sdk-kmp
Respository for utiliies for testing


CHANGELOG.adoc
= Changelog
:link-repository: https://github.com/d4l-data4life/hc-test-util-sdk-kmp
:doctype: article
:toc: macro
:toclevels: 1
:toc-title:
:icons: font
:imagesdir: assets/images
ifdef::env-github[]
:warning-caption: :warning:
:caution-caption: :fire:
:important-caption: :exclamation:
:note-caption: :paperclip:
:tip-caption: :bulb:
endif::[]

All notable changes to this project will be documented in this file.

The format is based on http://keepachangelog.com/en/1.0.0/[Keep a Changelog]
and this project adheres to http://semver.org/spec/v2.0.0.html[Semantic Versioning].

Template could be found here: link:https://github.com/d4l-data4life/hc-readme-template/blob/main/TEMPLATE_CHANGELOG.adoc[CHANGELOG]

[discrete]
==== Versions
toc::[]

== https://github.com/d4l-data4life/hc-test-util-sdk-kmp/compare/v0.4.1\...main[Unreleased]

=== Added

=== Changed

=== Deprecated

=== Removed

=== Fixed

=== Security

=== Bumped

== https://github.com/d4l-data4life/hc-test-util-sdk-kmp/compare/v0.4.1\...0.5.0[0.5.0]

Re-Release of 0.4.1 due to publishing issues

== https://github.com/d4l-data4life/hc-test-util-sdk-kmp/compare/v0.4.0\...0.4.1[0.4.1]

=== Bumped

* Robolectric 4.7.3 -> 4.8

=== Changed

* Android Target SDK 30 -> 31

== https://github.com/d4l-data4life/hc-test-util-sdk-kmp/compare/v0.3.0\...v0.4.0[0.4.0]

=== Bumped

* Gradle 7.2 -> 7.4.2
* Android Gradle Plugin 7.0.2 -> 7.1.3
* *_BREAKING_* Kotlin -> 1.6.10
* *_BREAKING_* Coroutines 1.5.1 -> 1.6.0
* Ktor 1.6.2 -> 1.6.8
* Desugar 1.0.4 -> 1.1.5
* Android Ktx 1.6.0 -> 1.7.0
* Android AppCompat 1.3.1 -> 1.4.1
* Android ConstraintLayout 2.0.1 -> 2.1.3
* Android Material 1.4.0 -> 1.5.0
* Robolectric 4.6.1 -> 4.7.3

== https://github.com/d4l-data4life/hc-test-util-sdk-kmp/compare/v0.2.0\...v0.3.0[0.3.0]

=== Added

* `AndroidOnly`, `JvmOnly` and `iOSOnly` annotations for common

=== Bumped

* Gradle 6.9 -> 7.2
* *_BREAKING_* Kotlin 1.4.32 -> 1.5.21
* *_BREAKING_* Android Gradle Plugin 4.2.2 -> 7.0.2
* *_BREAKING_* Coroutines 1.4.3 -> 1.5.1
* *_BREAKING_* Ktor 1.5.4 -> 1.6.2
* Android XKtx 1.3.1 -> 1.6.0
* Android AppCompat 1.2.0 -> 1.3.1
* Android Test 1.1.2 -> 1.4.0
* Android Espresso 3.3.0 -> 3.4.0
* Android Material Components 1.2.0 -> 1.4.0
* JUnit 4.13.1 -> 4.13.2

== https://github.com/d4l-data4life/hc-test-util-sdk-kmp/compare/v0.1.0\...v0.2.0[0.2.0]

=== Added

* RobolectricTestRunner for common

=== Bumped

* Robolectric 4.4 -> 4.6.1


== https://github.com/d4l-data4life/hc-test-util-sdk-kmp/compare/v0.1.0[0.1.0]

Initial release taken to be open sourced.


buildSrc/src/main/kotlin/care/data4life/sdk/util/test/LibraryConfig.kt
/*
 * Copyright (c) 2021 D4L data4life gGmbH / All rights reserved.
 *
 * D4L owns all legal rights, title and interest in and to the Software Development Kit ("SDK"),
 * including any intellectual property rights that subsist in the SDK.
 *
 * The SDK and its documentation may be accessed and used for viewing/review purposes only.
 * Any usage of the SDK for other purposes, including usage for the development of
 * applications/third-party applications shall require the conclusion of a license agreement
 * between you and D4L.
 *
 * If you are interested in licensing the SDK for your own applications/third-party
 * applications and/or if you’d like to contribute to the development of the SDK, please
 * contact D4L by email to help@data4life.care.
 */
package care.data4life.sdk.util.test

object LibraryConfig {
    val publishConfig = PublishConfig
    val android = AndroidLibraryConfig

    const val group = "care.data4life"
    const val name = "hc-test-util-sdk-kmp"

    const val githubOwner = "d4l-data4life"
    const val githubRepository = "hc-test-util-sdk-kmp"

    object PublishConfig {
        const val groupId = "care.data4life.hc-test-util-sdk-kmp"
        const val description = "Test-Util provide some convenience and necessary helpers in order to enable crossplatform testing."

        const val year = "2021"

        // URL
        const val host = "github.com"
        const val path = "$githubOwner/$githubRepository"

        const val url = "https://$host/$path"

        // DEVELOPER
        const val developerId = "d4l-data4life"
        const val developerName = "D4L data4life gGmbH"
        const val developerEmail = "mobile@data4life.care"

        // LICENSE
        const val licenseName = "Private"
        const val licenseUrl = "$url/blob/main/LICENSE"
        const val licenseDistribution = "repo"

        // SCM
        const val scmUrl = "git://$host/$path.git"
        const val scmConnection = "scm:$scmUrl"
        const val scmDeveloperConnection = scmConnection
    }

    object AndroidLibraryConfig {
        const val minSdkVersion = 23
        const val compileSdkVersion = 30
        const val targetSdkVersion = 30

        const val resourcePrefix = "hc_test_util_"
    }
}


test-util/src/iosMain/kotlin/care/data4life/sdk/util/test/NSDataMapper.kt
/*
 * Copyright (c) 2021 D4L data4life gGmbH / All rights reserved.
 *
 * D4L owns all legal rights, title and interest in and to the Software Development Kit ("SDK"),
 * including any intellectual property rights that subsist in the SDK.
 *
 * The SDK and its documentation may be accessed and used for viewing/review purposes only.
 * Any usage of the SDK for other purposes, including usage for the development of
 * applications/third-party applications shall require the conclusion of a license agreement
 * between you and D4L.
 *
 * If you are interested in licensing the SDK for your own applications/third-party
 * applications and/or if you’d like to contribute to the development of the SDK, please
 * contact D4L by email to help@data4life.care.
 */

package care.data4life.sdk.util.test

import kotlinx.cinterop.addressOf
import kotlinx.cinterop.allocArrayOf
import kotlinx.cinterop.memScoped
import kotlinx.cinterop.usePinned
import platform.Foundation.NSData
import platform.Foundation.create
import platform.posix.memcpy

internal object NSDataMapper {
    // see: https://github.com/JetBrains/kotlin-native/issues/3172
    // see: https://stackoverflow.com/questions/58521108/how-to-convert-kotlin-bytearray-to-nsdata-and-viceversa
    // see: https://gist.github.com/noahsark769/61cfb7a8b7231e2069a9dab94cf74a62
    fun toByteArray(data: NSData): ByteArray {
        return ByteArray(data.length.toInt()).apply {
            usePinned {
                memcpy(it.addressOf(0), data.bytes, data.length)
            }
        }
    }

    @ExperimentalUnsignedTypes
    fun toNSData(data: ByteArray): NSData {
        return memScoped {
            NSData.create(
                bytes = allocArrayOf(data),
                length = data.size.toULong()
            )
        }
    }
}


test-util/src/iosTest/kotlin/care/data4life/sdk/util/test/NSDataMapperTest.kt
/*
 * Copyright (c) 2021 D4L data4life gGmbH / All rights reserved.
 *
 * D4L owns all legal rights, title and interest in and to the Software Development Kit ("SDK"),
 * including any intellectual property rights that subsist in the SDK.
 *
 * The SDK and its documentation may be accessed and used for viewing/review purposes only.
 * Any usage of the SDK for other purposes, including usage for the development of
 * applications/third-party applications shall require the conclusion of a license agreement
 * between you and D4L.
 *
 * If you are interested in licensing the SDK for your own applications/third-party
 * applications and/or if you’d like to contribute to the development of the SDK, please
 * contact D4L by email to help@data4life.care.
 */

package care.data4life.sdk.util.test

import kotlinx.cinterop.memScoped
import platform.Foundation.NSData
import platform.Foundation.create
import kotlin.test.Test
import kotlin.test.assertTrue

class NSDataMapperTest {
    @ExperimentalUnsignedTypes
    @Test
    fun `Given a toNSData is called with ByteArray converts to NSData`() {
        // Given
        val payload = "ABC".encodeToByteArray()

        // When
        val result: Any = NSDataMapper.toNSData(payload)

        // Then
        assertTrue(result is NSData)
    }

    @Test
    fun `Given a toByteArray is called with NSData converts to ByteArray`() {
        // Given
        val payload = memScoped {
            NSData.create(
                base64EncodedString = "U3RyaW5nIHRvIGVuY29kZQ==",
                options = 0
            ) ?: throw NullPointerException("Out of memory.")
        }

        // When
        val result: Any = NSDataMapper.toByteArray(payload)

        // Then
        assertTrue(result is ByteArray)
    }

    @ExperimentalUnsignedTypes
    @Test
    fun `Given a toNSData is called with a ByteArray and the result calls toByteArray it transforms it to is orginal state`() {
        // Given
        val payload = "ABC".encodeToByteArray()

        // When
        val nsData = NSDataMapper.toNSData(payload)
        val byteArray = NSDataMapper.toByteArray(nsData)

        // Then
        assertTrue(payload.contentEquals(byteArray))
    }
}


buildSrc/src/main/kotlin/Versions.kt
/*
 * Copyright (c) 2021 D4L data4life gGmbH / All rights reserved.
 *
 * D4L owns all legal rights, title and interest in and to the Software Development Kit ("SDK"),
 * including any intellectual property rights that subsist in the SDK.
 *
 * The SDK and its documentation may be accessed and used for viewing/review purposes only.
 * Any usage of the SDK for other purposes, including usage for the development of
 * applications/third-party applications shall require the conclusion of a license agreement
 * between you and D4L.
 *
 * If you are interested in licensing the SDK for your own applications/third-party
 * applications and/or if you’d like to contribute to the development of the SDK, please
 * contact D4L by email to help@data4life.care.
 */

object Versions {

    object GradlePlugins {
        const val kotlin = Versions.kotlin
        const val android = "4.2.2"
    }

    // Kotlin
    // https://github.com/JetBrains/kotlin
    const val kotlin = "1.4.32"

    // https://github.com/Kotlin/kotlinx.coroutines
    const val kotlinCoroutines = "1.4.3-native-mt"

    // https://github.com/touchlab/Stately
    const val stately = "1.1.6"

    // Network
    // https://ktor.io/
    const val ktor = "1.5.4"

    // D4L
    // https://github.com/d4l-data4life/hc-objc-util-sdk-kmp
    const val sdkObjcUtil = "0.1.1"

    // Android
    // https://developer.android.com/studio/write/java8-support
    const val androidDesugar = "1.0.4"

    // AndroidX
    // https://developer.android.com/jetpack/androidx
    const val androidX = "1.1.0"
    const val androidXKtx = "1.3.1"
    const val androidXAppCompat = "1.2.0"

    const val androidXConstraintLayout = "2.0.1"

    // https://developer.android.com/testing
    const val androidXTest = "1.1.2"
    const val androidXEspresso = "3.3.0"
    const val androidXUiAutomator = "2.2.0"

    // Material
    /**
     * [Material Android](https://github.com/material-components/material-components-android)
     */
    const val material = "1.2.0"

    // Junit Test
    const val testJUnit = "4.13.1"

    /**
     * [Robolectric](https://github.com/robolectric/robolectric)
     */
    const val robolectric = "4.4"
}


buildSrc/src/main/kotlin/Dependencies.kt
/*
 * Copyright (c) 2021 D4L data4life gGmbH / All rights reserved.
 *
 * D4L owns all legal rights, title and interest in and to the Software Development Kit ("SDK"),
 * including any intellectual property rights that subsist in the SDK.
 *
 * The SDK and its documentation may be accessed and used for viewing/review purposes only.
 * Any usage of the SDK for other purposes, including usage for the development of
 * applications/third-party applications shall require the conclusion of a license agreement
 * between you and D4L.
 *
 * If you are interested in licensing the SDK for your own applications/third-party
 * applications and/or if you’d like to contribute to the development of the SDK, please
 * contact D4L by email to help@data4life.care.
 */

object Dependencies {

    val kotlin = Kotlin
    object Kotlin {
        const val stdLib = multiplatform.kotlin.stdlibJdk
        const val reflect = "org.jetbrains.kotlin:kotlin-reflect:${Versions.kotlin}"
    }

    val multiplatform = Multiplatform
    object Multiplatform {

        val kotlin = Kotlin
        object Kotlin {
            const val stdlibCommon = "org.jetbrains.kotlin:kotlin-stdlib-common:${Versions.kotlin}"
            const val stdlibJdk = "org.jetbrains.kotlin:kotlin-stdlib:${Versions.kotlin}"
            const val stdlibJdk8 = "org.jetbrains.kotlin:kotlin-stdlib-jdk8:${Versions.kotlin}"
            const val stdlibJs = "org.jetbrains.kotlin:kotlin-stdlib:${Versions.kotlin}"
            const val stdlibNative = "org.jetbrains.kotlin:kotlin-stdlib:${Versions.kotlin}"
            const val stdlibAndroid = "org.jetbrains.kotlin:kotlin-stdlib:${Versions.kotlin}"

            const val testCommon = "org.jetbrains.kotlin:kotlin-test-common:${Versions.kotlin}"
            const val testCommonAnnotations = "org.jetbrains.kotlin:kotlin-test-annotations-common:${Versions.kotlin}"
            const val testJvm = "org.jetbrains.kotlin:kotlin-test:${Versions.kotlin}"
            const val testJvmJunit = "org.jetbrains.kotlin:kotlin-test-junit:${Versions.kotlin}"
        }

        val coroutines = Coroutines

        object Coroutines {
            // https://github.com/Kotlin/kotlinx.coroutines
            const val common =
                "org.jetbrains.kotlinx:kotlinx-coroutines-core:${Versions.kotlinCoroutines}"
            const val android =
                "org.jetbrains.kotlinx:kotlinx-coroutines-android:${Versions.kotlinCoroutines}"
            const val test = "org.jetbrains.kotlinx:kotlinx-coroutines-test:${Versions.kotlinCoroutines}"

            const val stately = "co.touchlab:stately-common:${Versions.stately}"
        }

        val ktor = Ktor

        object Ktor {
            const val common = "io.ktor:ktor-client-core:${Versions.ktor}"
            const val mock = "io.ktor:ktor-client-mock:${Versions.ktor}"
        }

        val d4l = D4L

        object D4L {
            const val sdkObjcUtil = "care.data4life.hc-objc-util-sdk-kmp:util:${Versions.sdkObjcUtil}"
        }
    }

    val test = Test
    object Test {
        const val junit = "junit:junit:${Versions.testJUnit}"
    }

    val android = Android
    object Android {
        // Kotlin
        const val kotlinStdLib = "org.jetbrains.kotlin:kotlin-stdlib:${Versions.kotlin}"

        // Android
        const val desugar = "com.android.tools:desugar_jdk_libs:${Versions.androidDesugar}"

        val androidX = AndroidX
        object AndroidX {
            // AndroidX
            const val ktx = "androidx.core:core-ktx:${Versions.androidXKtx}"
            const val appCompat = "androidx.appcompat:appcompat:${Versions.androidXAppCompat}"
            const val constraintLayout = "androidx.constraintlayout:constraintlayout:${Versions.androidXConstraintLayout}"
        }

        // Material
        const val material = "com.google.android.material:material:${Versions.material}"

        // Test
        const val robolectric = "org.robolectric:robolectric:${Versions.robolectric}"
    }

    val androidTest = AndroidTest
    object AndroidTest {
        const val core = "androidx.test:core:${Versions.androidXTest}"
        const val runner = "androidx.test:runner:${Versions.androidXTest}"
        const val rules = "androidx.test:rules:${Versions.androidXTest}"

        const val junit = "androidx.test.ext:junit:${Versions.androidXTest}"

        const val espressoCore = "androidx.test.espresso:espresso-core:${Versions.androidXEspresso}"
        const val espressoIntents =
            "androidx.test.espresso:espresso-intents:${Versions.androidXEspresso}"
        const val espressoWeb = "androidx.test.espresso:espresso-web:${Versions.androidXEspresso}"

        const val uiAutomator =
            "androidx.test.uiautomator:uiautomator:${Versions.androidXUiAutomator}"
    }
}


buildSrc/src/main/kotlin/LibraryConfig.kt
/*
 * Copyright (c) 2021 D4L data4life gGmbH / All rights reserved.
 *
 * D4L owns all legal rights, title and interest in and to the Software Development Kit ("SDK"),
 * including any intellectual property rights that subsist in the SDK.
 *
 * The SDK and its documentation may be accessed and used for viewing/review purposes only.
 * Any usage of the SDK for other purposes, including usage for the development of
 * applications/third-party applications shall require the conclusion of a license agreement
 * between you and D4L.
 *
 * If you are interested in licensing the SDK for your own applications/third-party
 * applications and/or if you’d like to contribute to the development of the SDK, please
 * contact D4L by email to help@data4life.care.
 */

object LibraryConfig {
    const val group = "care.data4life"
    const val name = "hc-test-util-sdk-kmp"

    const val githubOwner = "d4l-data4life"
    const val githubRepository = "hc-test-util-sdk-kmp"

    val publish = PublishConfig

    object PublishConfig {
        const val name = LibraryConfig.name
        const val description = "Test-Util provide some convenience and necessary helpers in order to enable crossplatform testing."
        const val groupId = "${LibraryConfig.group}.${LibraryConfig.name}"

        const val year = "2020"

        // URL
        const val host = "github.com"
        const val path = "$githubOwner/$githubRepository"

        const val url = "https://$host/$path"

        // DEVELOPER
        const val developerId = "d4l-data4life"
        const val developerName = "D4L data4life gGmbH"
        const val developerEmail = "mobile@data4life.care"

        // LICENSE
        const val licenseName = "Private"
        const val licenseUrl = "$url/blob/main/LICENSE"
        const val licenseDistribution = "repo"

        // SCM
        const val scmUrl = "git://$host/$path.git"
        const val scmConnection = "scm:$scmUrl"
        const val scmDeveloperConnection = scmConnection
    }

    val android = AndroidLibraryConfig

    object AndroidLibraryConfig {
        const val minSdkVersion = 23
        const val compileSdkVersion = 30
        const val targetSdkVersion = 30

        const val resourcePrefix = "hc_test_util_"
    }
}
